File: packages/python/src/ai_kit/providers/gemini.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Dict, Iterable, List, Optional

from ..http import request_json, request_stream
from ..errors import ErrorKind, KitErrorPayload, InferenceKitError
from ..sse import iter_sse_events
from ..types import (
    GenerateInput,
    GenerateOutput,
    ImageGenerateInput,
    ImageGenerateOutput,
    ModelCapabilities,
    ModelMetadata,
    Provider,
    StreamChunk,
    ensure_messages,
)


@dataclass
class GeminiConfig:
    api_key: str = ""
    api_keys: Optional[List[str]] = None
    base_url: str = "https://generativelanguage.googleapis.com"
    timeout: Optional[float] = None


class GeminiAdapter:
    def __init__(self, config: GeminiConfig, provider: Provider = "google") -> None:
        self.config = config
        self.provider = provider
        self.base_url = config.base_url.rstrip("/")

    def list_models(self) -> List[ModelMetadata]:
        url = f"{self.base_url}/v1beta/models?key={self.config.api_key}"
        payload = request_json("GET", url, self._headers(), timeout=self.config.timeout)
        models: List[ModelMetadata] = []
        for model in payload.get("models", []) or []:
            name = model.get("name")
            if not name:
                continue
            model_id = name.replace("models/", "")
            models.append(
                ModelMetadata(
                    id=model_id,
                    displayName=model.get("displayName") or model_id,
                    provider=self.provider,
                    family=_derive_family(model_id),
                    capabilities=ModelCapabilities(
                        text=True,
                        vision=True,
                        tool_use=True,
                        structured_output=True,
                        reasoning=False,
                    ),
                    contextWindow=model.get("inputTokenLimit"),
                )
            )
        return models

    def generate(self, input: GenerateInput) -> GenerateOutput:
        url = f"{self.base_url}/v1beta/models/{_ensure_models_prefix(input.model)}:generateContent?key={self.config.api_key}"
        payload = request_json(
            "POST",
            url,
            self._headers(),
            json_body=_build_payload(input),
            timeout=self.config.timeout,
        )
        return _normalize_output(payload)

    def generate_image(self, input: ImageGenerateInput) -> ImageGenerateOutput:
        url = f"{self.base_url}/v1beta/models/{_ensure_models_prefix(input.model)}:generateContent?key={self.config.api_key}"
        payload = request_json(
            "POST",
            url,
            self._headers(),
            json_body=_build_image_payload(input),
            timeout=self.config.timeout,
        )
        inline = _extract_inline_image(payload)
        if not inline or not inline.get("data"):
            raise InferenceKitError(
                KitErrorPayload(
                    kind=ErrorKind.UNKNOWN,
                    message="Gemini image response missing inline data",
                    provider=self.provider,
                )
            )
        mime = inline.get("mimeType") or "image/png"
        return ImageGenerateOutput(mime=mime, data=inline["data"], raw=payload)

    def generate_mesh(self, input: "MeshGenerateInput"):
        raise InferenceKitError(
            KitErrorPayload(
                kind=ErrorKind.UNSUPPORTED,
                message="Gemini mesh generation is not supported",
                provider=self.provider,
            )
        )

    def stream_generate(self, input: GenerateInput) -> Iterable[StreamChunk]:
        url = f"{self.base_url}/v1beta/models/{_ensure_models_prefix(input.model)}:streamGenerateContent?alt=sse&key={self.config.api_key}"
        response = request_stream(
            "POST",
            url,
            self._headers(),
            json_body=_build_payload(input),
            timeout=self.config.timeout,
        )
        for event in iter_sse_events(response.iter_lines(decode_unicode=True)):
            data = event.get("data")
            if not data or data == "[DONE]":
                continue
            try:
                payload = json.loads(data)
            except json.JSONDecodeError:
                continue
            text = _extract_text(payload)
            if text:
                yield StreamChunk(type="delta", textDelta=text)
        response.close()

    def _headers(self) -> Dict[str, str]:
        return {"content-type": "application/json"}


def _build_payload(input: GenerateInput) -> Dict[str, Any]:
    messages = ensure_messages(input.messages)
    contents = []
    for message in messages:
        parts = []
        for part in message.get("content", []) or []:
            if part.get("type") == "text":
                parts.append({"text": part.get("text")})
        if parts:
            contents.append({"role": message.get("role"), "parts": parts})
    payload: Dict[str, Any] = {"contents": contents}
    generation_config: Dict[str, Any] = {}
    if input.temperature is not None:
        generation_config["temperature"] = input.temperature
    if input.topP is not None:
        generation_config["topP"] = input.topP
    if input.maxTokens is not None:
        generation_config["maxOutputTokens"] = input.maxTokens
    if generation_config:
        payload["generationConfig"] = generation_config
    return payload


def _build_image_payload(input: ImageGenerateInput) -> Dict[str, Any]:
    parts: List[Dict[str, Any]] = [{"text": input.prompt}]
    for image in input.inputImages or []:
        if image.base64:
            parts.append(
                {
                    "inlineData": {
                        "mimeType": image.mediaType or "image/png",
                        "data": image.base64,
                    }
                }
            )
        elif image.url:
            parts.append({"fileData": {"fileUri": image.url}})
    return {"contents": [{"role": "user", "parts": parts}]}


def _normalize_output(payload: Dict[str, Any]) -> GenerateOutput:
    text = _extract_text(payload)
    return GenerateOutput(text=text or None, raw=payload)


def _extract_text(payload: Dict[str, Any]) -> str:
    candidates = payload.get("candidates", []) or []
    if not candidates:
        return ""
    content = candidates[0].get("content", {}) or {}
    parts = content.get("parts", []) or []
    texts = [part.get("text", "") for part in parts if part.get("text")]
    return "".join(texts)


def _extract_inline_image(payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    candidates = payload.get("candidates", []) or []
    for candidate in candidates:
        content = candidate.get("content", {}) or {}
        parts = content.get("parts", []) or []
        for part in parts:
            inline = part.get("inlineData")
            if isinstance(inline, dict) and inline.get("data"):
                return inline
    return None


def _ensure_models_prefix(model_id: str) -> str:
    if model_id.startswith("models/"):
        return model_id
    return f"models/{model_id}"


def _derive_family(model_id: str) -> str:
    parts = model_id.split("-")
    return "-".join(parts[:2]) if len(parts) >= 2 else model_id

File: packages/node/src/adapters/gemini.ts
import { ProviderAdapter } from "../core/provider.js";
import {
  FetchLike,
  GenerateInput,
  GenerateOutput,
  GoogleProviderConfig,
  ImageGenerateInput,
  ImageGenerateOutput,
  ModelMetadata,
  Provider,
  ResponseFormat,
  StreamChunk,
  ToolCall,
  ToolDefinition,
  Usage,
} from "../core/types.js";
import { AiKitError } from "../core/errors.js";
import { ErrorKind } from "../core/types.js";
import { parseEventData, streamSSE } from "../core/stream.js";

interface GeminiModelList {
  models: Array<{
    name: string;
    displayName: string;
    description?: string;
    supportedGenerationMethods?: string[];
    inputTokenLimit?: number;
    outputTokenLimit?: number;
  }>;
}

interface GeminiContent {
  role: string;
  parts: Array<GeminiPart>;
}

type GeminiPart =
  | { text: string }
  | { inlineData: { mimeType: string; data: string } }
  | { fileData: { mimeType?: string; fileUri: string } }
  | {
      functionCall: {
        name: string;
        args?: Record<string, unknown>;
      };
    }
  | {
      functionResponse: {
        name: string;
        response?: unknown;
      };
    };

interface GeminiGenerateResponse {
  candidates?: Array<{
    content?: {
      parts?: GeminiPart[];
    };
    finishReason?: string;
  }>;
  usageMetadata?: {
    promptTokenCount?: number;
    candidatesTokenCount?: number;
    totalTokenCount?: number;
  };
}

interface GeminiStreamEvent {
  candidates?: GeminiGenerateResponse["candidates"];
  usageMetadata?: GeminiGenerateResponse["usageMetadata"];
}

export class GoogleAdapter implements ProviderAdapter {
  readonly provider = Provider.Google;
  private readonly baseURL: string;
  private readonly apiKey: string;
  private readonly fetchImpl?: FetchLike;

  constructor(config: GoogleProviderConfig, fetchImpl?: FetchLike) {
    this.baseURL =
      config.baseURL ?? "https://generativelanguage.googleapis.com";
    this.apiKey = config.apiKey;
    this.fetchImpl = fetchImpl;
  }

  async listModels(): Promise<ModelMetadata[]> {
    const data = await this.fetchJSON<GeminiModelList>(`/v1beta/models`, {
      method: "GET",
    });
    return data.models.map((model) => {
      const id = model.name.replace("models/", "");
      return {
        id,
        displayName: model.displayName ?? id,
        provider: this.provider,
        family: id.split("-").slice(0, 3).join("-"),
        capabilities: {
          text: true,
          vision: true,
          tool_use: true,
          structured_output: true,
          reasoning: false,
        },
        contextWindow: model.inputTokenLimit,
      };
    });
  }

  async generate(input: GenerateInput): Promise<GenerateOutput> {
    const response = await this.fetchJSON<GeminiGenerateResponse>(
      this.buildModelPath(input.model, ":generateContent"),
      {
        method: "POST",
        body: this.buildPayload({ ...input, stream: false }),
        signal: input.signal,
      },
    );
    return this.normalizeResponse(response);
  }

  async generateImage(
    input: ImageGenerateInput,
  ): Promise<ImageGenerateOutput> {
    const payload = buildImagePayload(input);
    const response = await this.fetchJSON<GeminiGenerateResponse>(
      this.buildModelPath(input.model, ":generateContent"),
      {
        method: "POST",
        body: payload,
      },
    );
    const image = extractInlineImage(response);
    if (!image) {
      throw new AiKitError({
        kind: ErrorKind.Unknown,
        message: "Gemini image response missing inline data",
        provider: this.provider,
      });
    }
    return {
      mime: image.mimeType ?? "image/png",
      data: image.data,
      raw: response,
    };
  }

  streamGenerate(input: GenerateInput): AsyncIterable<StreamChunk> {
    const body = this.buildPayload({ ...input, stream: true });
    return this.streamContent(
      this.buildModelPath(input.model, ":streamGenerateContent"),
      body,
      input.signal,
    );
  }

  private buildPayload(input: GenerateInput & { stream?: boolean }) {
    const { systemInstruction, contents } = buildGeminiContents(input.messages);
    const generationConfig = buildGenerationConfig(
      input.responseFormat,
      input.temperature,
      input.topP,
      input.maxTokens,
    );
    return {
      contents,
      systemInstruction,
      generationConfig,
      tools: buildTools(input.tools),
      toolConfig: buildToolConfig(input.toolChoice),
      safetySettings: undefined,
      stream: input.stream ?? false,
    };
  }

  private normalizeResponse(response: GeminiGenerateResponse): GenerateOutput {
    const candidate = response.candidates?.[0];
    const parts = candidate?.content?.parts ?? [];
    const text = parts
      .filter((part): part is { text: string } => "text" in part)
      .map((part) => part.text)
      .join("");
    const toolCalls = parts
      .filter((part): part is { functionCall: { name: string; args?: Record<string, unknown> } } =>
        Boolean((part as any).functionCall),
      )
      .map((part, index) => ({
        id: part.functionCall.name ?? `call_${index}`,
        name: part.functionCall.name ?? `call_${index}`,
        argumentsJson: JSON.stringify(part.functionCall.args ?? {}),
      }));
    return {
      text: text || undefined,
      toolCalls: toolCalls.length ? toolCalls : undefined,
      finishReason: candidate?.finishReason ?? "stop",
      usage: mapGeminiUsage(response.usageMetadata),
      raw: response,
    };
  }

  private async *streamContent(
    path: string,
    body: unknown,
    signal?: AbortSignal,
  ): AsyncIterable<StreamChunk> {
    const response = await this.fetchRaw(`${path}?alt=sse`, {
      method: "POST",
      body: JSON.stringify(body),
      signal,
      headers: {
        "content-type": "application/json",
      },
    });
    let usage: Usage | undefined;
    for await (const event of streamSSE(response)) {
      if (!event.data || event.data === "[DONE]") {
        continue;
      }
      const payload = parseEventData<GeminiStreamEvent>(event);
      if (!payload) continue;
      if (payload.usageMetadata) {
        usage = mapGeminiUsage(payload.usageMetadata);
      }
      for (const candidate of payload.candidates ?? []) {
        const parts = candidate?.content?.parts ?? [];
        for (const part of parts) {
          if ("text" in part && part.text) {
            yield { type: "delta", textDelta: part.text };
          } else if ("functionCall" in part) {
            yield {
              type: "tool_call",
              call: {
                id: part.functionCall.name,
                name: part.functionCall.name,
                argumentsJson: JSON.stringify(part.functionCall.args ?? {}),
              },
            };
          }
        }
      }
    }
    yield {
      type: "message_end",
      usage,
      finishReason: "stop",
    };
  }

  private buildModelPath(model: string, method: string) {
    const normalized = model.startsWith("models/") ? model : `models/${model}`;
    return `/v1beta/${normalized}${method}`;
  }

  private async fetchJSON<T>(
    path: string,
    init: RequestInit & { body?: Record<string, unknown> },
  ): Promise<T> {
    const response = await this.fetchRaw(path, {
      ...init,
      body:
        init.body && typeof init.body !== "string"
          ? JSON.stringify(init.body)
          : init.body,
      headers: {
        "content-type": "application/json",
        ...(init.headers ?? {}),
      },
    });
    return (await response.json()) as T;
  }

  private async fetchRaw(path: string, init: RequestInit): Promise<Response> {
    const fetchImpl = this.fetchImpl ?? globalThis.fetch;
    if (!fetchImpl) {
      throw new AiKitError({
        kind: ErrorKind.Unsupported,
        message: "global fetch is not available",
        provider: this.provider,
      });
    }
    const separator = path.includes("?") ? "&" : "?";
    const url = `${this.baseURL}${path}${separator}key=${encodeURIComponent(this.apiKey)}`;
    const response = await fetchImpl(url, init);
    if (!response.ok) {
      const text = await response.text().catch(() => "");
      throw new AiKitError({
        kind:
          response.status === 401 || response.status === 403
            ? ErrorKind.ProviderAuth
            : response.status === 429
              ? ErrorKind.ProviderRateLimit
              : response.status >= 500
                ? ErrorKind.ProviderUnavailable
                : ErrorKind.Unknown,
        message: text || `Gemini error (${response.status})`,
        provider: this.provider,
        upstreamStatus: response.status,
      });
    }
    return response;
  }
}

function buildGeminiContents(messages: GenerateInput["messages"]) {
  const systemParts: string[] = [];
  const contents: GeminiContent[] = [];
  for (const message of messages) {
    if (message.role === "system") {
      systemParts.push(
        message.content
          .map((part) => (part.type === "text" ? part.text : ""))
          .join(""),
      );
      continue;
    }
    if (message.role === "tool") {
      contents.push({
        role: "user",
        parts: message.content
          .filter((part) => part.type === "text")
          .map((part) => ({
            functionResponse: {
              name: message.name ?? message.toolCallId ?? "tool",
              response: safeParseJSON(part.text),
            },
          })),
      });
      continue;
    }
    contents.push({
      role: message.role === "assistant" ? "model" : "user",
      parts: message.content
        .filter((part) => part.type !== "tool_use")
        .map((part) => {
        if (part.type === "text") {
          return { text: part.text };
        }
        if (part.image.base64) {
          return {
            inlineData: {
              mimeType: part.image.mediaType ?? "image/png",
              data: part.image.base64,
            },
          };
        }
        return {
          fileData: {
            mimeType: part.image.mediaType,
            fileUri: part.image.url ?? "",
          },
        };
      }),
    });
  }
  return {
    systemInstruction: systemParts.length
      ? {
          role: "system",
          parts: systemParts.map((text) => ({ text })),
        }
      : undefined,
    contents,
  };
}

function buildImagePayload(input: ImageGenerateInput) {
  const parts: GeminiPart[] = [{ text: input.prompt }];
  for (const image of input.inputImages ?? []) {
    if (image.base64) {
      parts.push({
        inlineData: {
          mimeType: image.mediaType ?? "image/png",
          data: image.base64,
        },
      });
    } else if (image.url) {
      parts.push({
        fileData: {
          mimeType: image.mediaType,
          fileUri: image.url,
        },
      });
    }
  }
  return {
    contents: [{ role: "user", parts }],
  };
}

function extractInlineImage(response: GeminiGenerateResponse) {
  for (const candidate of response.candidates ?? []) {
    for (const part of candidate.content?.parts ?? []) {
      if ("inlineData" in part && part.inlineData?.data) {
        return part.inlineData;
      }
    }
  }
  return null;
}

function buildGenerationConfig(
  format: ResponseFormat | undefined,
  temperature?: number,
  topP?: number,
  maxTokens?: number,
) {
  const config: Record<string, unknown> = {};
  if (temperature !== undefined) config.temperature = temperature;
  if (topP !== undefined) config.topP = topP;
  if (maxTokens !== undefined) config.maxOutputTokens = maxTokens;
  if (format?.type === "json_schema") {
    config.responseMimeType = "application/json";
    config.responseSchema = format.jsonSchema.schema;
  }
  return Object.keys(config).length ? config : undefined;
}

function buildTools(tools?: ToolDefinition[]) {
  if (!tools?.length) return undefined;
  return [
    {
      functionDeclarations: tools.map((tool) => ({
        name: tool.name,
        description: tool.description,
        parameters: tool.parameters,
      })),
    },
  ];
}

function buildToolConfig(choice?: GenerateInput["toolChoice"]) {
  if (!choice) return undefined;
  if (choice === "auto") {
    return {
      functionCallConfig: { mode: "AUTO" },
    };
  }
  if (choice === "none") {
    return {
      functionCallConfig: { mode: "NONE" },
    };
  }
  return {
    functionCallConfig: {
      mode: "ANY",
      allowedFunctionNames: [choice.name],
    },
  };
}

function mapGeminiUsage(meta?: GeminiGenerateResponse["usageMetadata"]) {
  if (!meta) return undefined;
  return {
    inputTokens: meta.promptTokenCount,
    outputTokens: meta.candidatesTokenCount,
    totalTokens: meta.totalTokenCount,
  };
}

function safeParseJSON(value: string) {
  try {
    return JSON.parse(value);
  } catch {
    return value;
  }
}
